File: 20-valid-parentheses/20-valid-parentheses.js
let Stack = function(){
    this.innerStack = []
    this.push = (val)=>{
        this.innerStack.push(val)
    }
    this.pop = ()=>{
        const popped = this.innerStack.splice(this.innerStack.length-1)
        return popped[0]
    }
    this.getLength = ()=>{
        return this.innerStack.length
    }
}

/**
 * @param {string} s
 * @return {boolean}
 */
var isValid = function(s) {
    const parenthStack = new Stack()
    const lookUpDict = {
        '}':'{',
        ')':'(',
        ']':'['
    }
    for(i in s){
        if(s[i] === '{' || s[i] === '[' || s[i] === '('){
            parenthStack.push(s[i])
        }else{
            let popped = parenthStack.pop();
            if(popped !== lookUpDict[s[i]]){
                return false
            }
        }
    }
    
    if(parenthStack.getLength() === 0){
        return true
    }
    return false
};

// learned I can just push the closing parentheses into the stack rather than looking it up from a map or dictionary all the time

File: 20-valid-parentheses/20-valid-parentheses.cs
public class Solution {
    public bool IsValid(string s) {
        Stack<char> bracketStack = new Stack<char>();
        Dictionary<char, char> lookUpDict = new Dictionary<char, char>();
        lookUpDict.Add('}','{');
        lookUpDict.Add(')','(');
        lookUpDict.Add(']','[');
        
        foreach(char c in s){
            if (lookUpDict.ContainsKey(c)){
                if(bracketStack.Count>0){
                   char popped = bracketStack.Pop();
                    if (popped != lookUpDict[c]){
                        return false;
                    }else{
                        continue;
                    } 
                }else{
                    return false;
                }
                
            }else{
                bracketStack.Push(c);
            }
                
        }
        if(bracketStack.Count == 0){
            return true;
        }else{
            return false;
        }
        
    }
}